agent/taskresource/envFiles/envfile.go

```diff
@@ -0,0 +1,198 @@
+	if !ok {
+		err := errors.New("environment file resource: unable to find execution role credentials")
+		envfile.setTerminalReason(err.Error())
+		return err
+	}
+
+	for _, envfileLocation := range envfile.envfileLocations {
+		// if we support types besides S3 ARN, we will need to add filtering before the below method is called
+
+		err := envfile.downloadEnvfileFromS3(envfileLocation.Value, executionCredentials)
+		if err != nil {
+			err = errors.Wrapf(err, "unable to download envfile with ARN %v from s3", envfileLocation.Value)
+			envfile.setTerminalReason(err.Error())
+			return err
+		}
+	}
+
+	return nil
+}
+
+func (envfile *EnvironmentFileResource) downloadEnvfileFromS3(envFilePath string, executionCredentials credentials.TaskIAMRoleCredentials) error {
+	bucket, key, err := s3.ParseS3ARN(envFilePath)
+	if err != nil {
+		return errors.Wrapf(err, "unable to parse bucket and key from s3 ARN specified in environmentFile %s", envFilePath)
+	}
+
+	s3Client, err := envfile.s3ClientCreator.NewS3ClientForBucket(bucket, envfile.region, executionCredentials.GetIAMRoleCredentials())
+	if err != nil {
+		return errors.Wrapf(err, "unable to initialize s3 client for bucket %s", bucket)
+	}
+
+	seelog.Debugf("Downlading envfile with bucket name %v and key name %v", bucket, key)
+	// we save envfiles to path: /var/lib/ecs/data/envfiles/cluster_name/task_id/${s3filename.env}
+	downloadPath := filepath.Join(envfile.resourceDir, key)
+	err = envfile.writeEnvFile(func(file oswrapper.File) error {
+		return s3.DownloadFile(bucket, key, s3DownloadTimeout, file, s3Client)
+	}, downloadPath)
+
+	if err != nil {
+		return errors.Wrapf(err, "unable to download env file with key %s from bucket %s", key, bucket)
+	}
+
+	seelog.Debugf("Downloaded envfile from s3 and saved to %s", downloadPath)
+	return nil
+}
+
+func (envfile *EnvironmentFileResource) writeEnvFile(writeFunc func(file oswrapper.File) error, fullPathName string) error {
+	// File moves (renaming) are atomic while file writes are not
+	// so we write to a temp file before renaming to actual file
+	// multiple programs calling TempFile will not reference the same file
+	// so this should be ok to be called by multiple go routines
+	tmpFile, err := envfile.ioutil.TempFile(envfile.resourceDir, envTempFilePrefix)
+	if err != nil {
+		seelog.Infof("something went wrong trying to create a temp file with prefix", envTempFilePrefix)
+		return err
+	}
+	defer tmpFile.Close()
+
+	err = writeFunc(tmpFile)
+	if err != nil {
+		seelog.Infof("something went wrong trying to write to tmpFile %s", tmpFile.Name())
+		return err
+	}
+
+	// persist file to disk
+	err = tmpFile.Sync()
+	if err != nil {
+		seelog.Infof("something went wrong trying to persist envfile to disk")
+		return err
+	}
+
+	err = envfile.os.Rename(tmpFile.Name(), fullPathName)
+	if err != nil {
+		seelog.Infof("something went wrong when trying to rename envfile from %s to %s", tmpFile.Name(), fullPathName)
+		return err
+	}
+
```

Comment:
Are we going to have all other necessary functions that needed for task resource in later cr? Like Cleanup function, marshal/unmarshal Json function, etc. This file so far lacks many essential functions for task resource to work correctly.

Reply:
yeah i was going to include those when piping this resource through as part of task cycle

Reply:
nvm i'll add them in now

Reply:
added these, going to put accompanying unit tests in next PR